MRC-6040: Import display name from orderly custom metadata, store it in DB, and display it

david-mears-2 opened the pull request:
In this branch there's a new packet group in `/demos`, since none of the existing ones were making use of the `description.display_name` orderly property.

That property is referenced during the scheduled job that imports packets into the proxy db, so that the db column 'display_name' is now being used for the first time rather than just duplicating the name. When there's no display name, the column defaults to the name.

Then, the home page (the packet group index page) is now able to display the display name for packet groups, which is found by the API looking up the latest display name at the same time as it looks up the latest ID and latest time. The api also changes to allow you to filter by display_name, not just by name, from this page. The display name is also displayed on the packet group summary page (PacketGroup.tsx) and on the packet page and subpages (PacketHeader.tsx). In all those places, the component will render a little differently depending on whether the display name differs from the name.

I also introduce displaying the `description.long` orderly property (a lengthy description) in two places (the packet group page PacketGroup.tsx, and the packet summary page PacketDetails.tsx). This does not use the db, it uses a request for packet metadata.

# Screenshots

## Home page showing packet groups with and without display names

![Screenshot from 2024-12-04 17-04-48](https://github.com/user-attachments/assets/1d52c864-240f-4a7a-b1e6-b272c4b23c2a)

## Home page showing filtering on name

![Screenshot from 2024-12-04 17-04-58](https://github.com/user-attachments/assets/dea20f21-8f51-4ea8-bfcd-178f0c299784)

## Home page showing filtering on display name

![Screenshot from 2024-12-04 17-05-08](https://github.com/user-attachments/assets/6efbcbf8-9501-4969-8606-e562a7c64a80)

## Packet group page for packets that have display names and descriptions

![Screenshot from 2024-12-04 17-05-21](https://github.com/user-attachments/assets/18a52419-391b-4c57-804d-e9654bd14bf0)

## Packet group page for packets that have neither a display name nor a description

![Screenshot from 2024-12-04 17-06-33](https://github.com/user-attachments/assets/ba3ecb7a-3d5a-4ac7-b359-4b7a920a2108)

## Packet header (and summary page with description) for packets with display names and descriptions

![Screenshot from 2024-12-04 17-07-36](https://github.com/user-attachments/assets/988775c3-af6c-44b4-b459-f6fb43dbd83d)

## Packet header (and summary page with description) for packets with neither a display names nor a description (no change)

![Screenshot from 2024-12-04 17-07-50](https://github.com/user-attachments/assets/d8832fa7-6fb4-47d9-84ad-496ab23fe066)

## NB

I have *not* made changes to the runner logs page, which is shown below, as it would be fairly in depth (requires either another request to get the packet metadata from outpack server, or a db query to get the packet display name) and I'm not sure it's needed. But I could do it in another PR if required.

![image](https://github.com/user-attachments/assets/f4fc3ff5-1219-4eca-83f1-749b6635d2a5)

EmmaLRussell commented on api/app/src/main/kotlin/packit/service/PacketService.kt:
```diff
@@ -44,14 +45,21 @@ class BasePacketService(
     private val outpackServerClient: OutpackServer
 ) : PacketService
 {
+    private fun getDisplayNameForPacket(packet: OutpackMetadata): String {
+        val metadata = outpackServerClient.getMetadataById(packet.id)
+        val orderlyMetadata = metadata?.custom?.get("orderly") as? Map<*, *>
+        val description = orderlyMetadata?.get("description") as? Map<*, *>
+        return description?.get("display") as? String ?: packet.name
+    }
+
```
I'm surprised you need to do this - is the custom metadata not already in the metadata being returned for all new packets in line 59? If it's not it should be. Maybe it is but those values ujust aren't being passed out by `getMetadata`?  We shouldn't need to do a new call to the client per packet.

david-mears-2 replied:
It does seem inefficient doesn't it. Line 59 corresponds to calling `/packit/metadata` on the outpack server, which returns only id, name, parameters, and time, contrary to the documentation of outpack server: https://github.com/mrc-ide/outpack_server/pull/68

david-mears-2 replied:
So we'll change the outpack server

david-mears-2 replied:
Changed outpack server in https://github.com/mrc-ide/outpack_server/pull/69

EmmaLRussell commented on app/src/app/components/contents/PacketGroup/PacketTable.tsx:
```diff
@@ -15,6 +15,7 @@ export const PacketTable = () => {
   const { packetName } = useParams();
   const [pageNumber, setPageNumber] = useState(0);
 
+  // Rename useGetPacketGroups to useGetPacketGroupPages, and packetGroups to packetGroupPages?
```
Why?

david-mears-2 replied:
When you iterate over those things, the elements aren't packet groups, they are pages of packet groups. So later in the file we have eg 'packetGroups.first' which looks like it means 'the first packet group' but actually means 'the first page of packet groups'.

david-mears-2 replied:
Oh, I mistook this file for being `PacketGroupSummaryList.tsx`. In fact, each item returned by useGetPacketGroups (each so-called packetGroup) is a _page of packets_, not a packet group nor a page of packet groups. So these things should be renamed to useGetPacketPages and packetPages respectively.

david-mears-2 replied:
or more simply, `useGetPackets` and `pages`

david-mears-2 commented:
I wonder if we have any CI checks designed to notice when test coverage is too low -- I know my last commit is changing a lot of things that demand new unit tests!

david-mears-2 commented on api/app/src/main/kotlin/packit/controllers/PacketController.kt:
```diff
@@ -41,15 +41,15 @@ class PacketController(private val packetService: PacketService)
 
-    @GetMapping("/packetGroupSummary")
-    fun getPacketGroupSummary(
+    @GetMapping("/packetGroupSummaries")
```
This is one of a few renamings i did for increased precision. The endpoint returns a summary for each packet group (or, if a name to filter on is provided, a summary for each _matching_ packet group).

david-mears-2 commented on api/app/src/test/kotlin/packit/integration/repository/PacketRepositoryTest.kt:
```diff
@@ -119,38 +119,41 @@ class PacketRepositoryTest : RepositoryTest()
     @WithAuthenticatedUser(authorities = ["packet.read"])
 
-    fun `can get right order and data expected from findPacketGroupSummaryByName`()
+    fun `can get right order and data expected from findPacketGroupSummaryBySearchString`()
```
Note to self should say get not find

david-mears-2 commented on api/app/src/test/kotlin/packit/integration/repository/PacketRepositoryTest.kt:
```diff
@@ -159,6 +162,32 @@ class PacketRepositoryTest : RepositoryTest()
+
+    @Test
+    @WithAuthenticatedUser(authorities = ["packet.read"])
+    fun `can filter correctly when calling findPacketGroupSummaryBySearchString with part of a packet display name`()
```
Note to self fix name

david-mears-2 commented on app/src/app/components/contents/PacketGroup/PacketTable.tsx:
```diff
@@ -4,18 +4,17 @@ import { Skeleton } from "../../Base/Skeleton";
 import { ErrorComponent } from "../common/ErrorComponent";
 import { Pagination } from "../common/Pagination";
-import { useGetPacketGroups } from "./hooks/useGetPacketGroups";
+import { useGetPacketPages } from "./hooks/useGetPacketPages";
```
Another renaming: what’s returned is a paginated list of packets, not a packet group.

EmmaLRussell replied:
But what we're showing here is a single page of packet groups, isn't it? It seems odd to send "packetPages.content" into the data table as if it's going to display multiple pages...?
Would it be better as packetPage singular?

david-mears-2 commented on app/src/app/components/contents/PacketGroup/hooks/useGetPacketPages.ts:
```diff
@@ -3,14 +3,16 @@ import appConfig from "../../../../../config/appConfig";
 import { PageablePackets } from "../../../../../types";
 
-export const useGetPacketGroups = (packetName: string | undefined, pageNumber: number, pageSize: number) => {
+export const useGetPacketPages = (packetGroupName: string | undefined, pageNumber: number, pageSize: number) => {
```
Note to self, file name is not in sync with function name. In the back end this queries on packets so function rename is questionable.

david-mears-2 commented on app/src/app/components/routes/Router.tsx:
```diff
@@ -40,7 +40,7 @@ export function Router() {
             {/* <Route path="run-workflow" element={<WorkflowRunner />} /> */}
             {/* <Route path="documentation" element={<ProjectDocumentation />} /> */}
-            <Route path="/:packetName" element={<PacketGroup />} />
+            <Route path="/:packetGroupName" element={<PacketGroup />} />
```
Questionable

david-mears-2 commented on app/tailwind.config.js:
```diff
@@ -56,6 +56,9 @@ export const theme = {
       sans: ["Graphik", "sans-serif"],
       serif: ["Merriweather", "serif"]
     },
+    fontSize: {
```
Remove, not used

absternator commented on api/app/src/main/kotlin/packit/service/PacketService.kt:
```diff
@@ -44,14 +45,33 @@ class BasePacketService(
+        val orderlyMetadata = packet.custom?.get("orderly") as? Map<*, *>
+        val description = orderlyMetadata?.get("description") as? Map<*, *>
+        val orderlyDisplayName = description?.get("display") as? String
+        return if (!orderlyDisplayName.isNullOrBlank()) {
```
can use kotlin when here for concise 
```suggestion
       return when {
        !orderlyDisplayName.isNullOrBlank() -> orderlyDisplayName
        packet.custom?.get("display") is String -> packet.custom["display"] as String
        packet.custom?.get("display_name") is String -> packet.custom["display_name"] as String
        else -> packet.name
    }
```

absternator commented on api/app/src/main/kotlin/packit/service/PacketService.kt:
```diff
@@ -44,14 +45,33 @@ class BasePacketService(
+    // a display name.
+    // Also check for 'display name' keys that may exist in non-orderly outpack custom schemas.
+    // Falls back to name if no display name.
+    private fun getDisplayNameForPacket(packet: OutpackMetadata): String {
```
also didnt see unit tests for this method..... change to internal to be able to test and expose

david-mears-2 replied:
It's tested by the tests of importPackets. I think private functions are allowed, and that they shouldn't be separately unit tested.

david-mears-2 replied:
Because I should test the public interface ('guaranteeing' some level of stability for it), not the implementation details (which should be allowed to be changeable / not stable)

david-mears-2 replied:
```kt
val expectedDisplayNames = mapOf(
            "20240101-090000-4321gaga" to "test", // This packet has no display name, so should fall back to its name.
            "20190203-120000-1234dada" to "test name (latest display name)",
            "20190403-120000-1234dfdf" to "test2 name"
        )
```

That's the bit that tests the 'fall back' condition.

I don't bother testing for the non-orderly (outpack) conditions that speculatively check for keys called 'display', but could be persuaded to, just makes the test suite more verbose.

absternator commented on app/src/app/components/contents/PacketGroup/PacketGroup.tsx:
```diff
@@ -1,19 +1,48 @@
-import { useParams } from "react-router-dom";
 import { PacketTable } from "./PacketTable";
+import {useParams} from "react-router-dom";
```
formatting issues? there is no space b/w word and brackets

david-mears-2 replied:
Which formatter is meant to catch this?

EmmaLRussell replied:
I'd have thought prettier would have picked that up.

david-mears-2 replied:
Cool, I enabled prettier in my code editor and it picked this up.

absternator commented on app/src/app/components/contents/PacketGroup/PacketGroup.tsx:
```diff
@@ -1,19 +1,48 @@
 
 export const PacketGroup = () => {
   const { packetName } = useParams();
+  const { packetGroupSummaries, error: packetGroupError } =
```
there are now 3 network requests here including in Packettable... surely it can all be combined into 1... does pulling the one from packettable work? that should get you all the data you need

david-mears-2 replied:
I ended up implementing a new endpoint that would give us everything needed by the PacketGroup component: given a packet group name, return the display name and the latest packet's description. So now this only makes one network call! (Except PacketTable still makes a separate call for different info)

absternator commented on app/src/app/components/contents/PacketGroup/PacketGroup.tsx:
```diff
@@ -1,19 +1,48 @@
   const { packetName } = useParams();
+  const { packetGroupSummaries, error: packetGroupError } =
+    useGetPacketGroupSummaries(0, 1, packetName as string);
+  const latestPacketId = packetGroupSummaries?.content[0]?.latestId;
```
also just realised since display name is packet specific it doesn't make sense to show it at the top? since not all packets will have the same display name?

EmmaLRussell replied:
We're going to show the display name of the most recent packet.

absternator commented on app/src/app/components/contents/PacketGroup/PacketGroup.tsx:
```diff
@@ -1,19 +1,48 @@
+  const latestDescription = latestPacket?.custom?.orderly.description.long;
+
+  if ([packetError?.status, packetGroupError?.status].includes(HttpStatus.Unauthorized)) return <Unauthorized />;
+  if (packetGroupError || !packetGroupDisplayName) {
```
these get  cleaned up when using just 1 API call....

absternator commented on app/src/app/components/contents/PacketGroup/hooks/useGetPacketPages.ts:
```diff
@@ -3,14 +3,16 @@ import appConfig from "../../../../../config/appConfig";
+
   const { data, isLoading, error } = useSWR<PageablePackets>(
-    `${appConfig.apiUrl()}/packets/${packetName}?pageNumber=${pageNumber}&pageSize=${pageSize}`,
+    packetName ? url : null,
```
nice good catch!!!

absternator commented on app/src/app/components/contents/common/hooks/useGetPacketById.ts:
```diff
@@ -5,12 +5,17 @@ import { PacketMetadata } from "../../../../../types";
 export const useGetPacketById = (packetId: string | undefined) => {
   const { data, isLoading, error } = useSWR<PacketMetadata>(
-    `${appConfig.apiUrl()}/packets/metadata/${packetId}`,
+    packetId ?
```
good catch nice again

absternator commented on app/src/app/components/contents/common/hooks/useGetPacketById.ts:
```diff
@@ -5,12 +5,17 @@ import { PacketMetadata } from "../../../../../types";
 
+  const displayName = data?.custom?.orderly?.description?.display;
+
+  const packet = displayName ? { ...data, displayName } : data;
```
there is displayName in the interface? whats that for?
.. also instead of spreading you can probs just do `{data, displayName: data?.custom?.orderly?.description?.display;` ... it will be undefined anyways If not there

david-mears-2 replied:
That change makes sense.

Yes, in the PacketMetadata interface, there is a displayName property. You added it in November 2023! And now we're using it for the first time, in three components, all of which use it to pass it as a prop to the PacketHeader.

david-mears-2 replied:
Actually no sorry you didn't add it, you just changed the indentation... it was added here https://github.com/mrc-ide/packit/commit/fc88d6393b336908f848cd63b1aa096371d42ad4

absternator commented on app/src/app/components/contents/packets/PacketHeader.tsx:
```diff
@@ -1,13 +1,27 @@
-    </div>
-  );
-}
+export default function PacketHeader({ displayName, packetName, packetId }: PacketHeaderProps) {
```
possoble to make displayName to allow undefined as well... saves you having to do displayName ?? "" everywhere.. and the check can just be if displayName

david-mears-2 replied:
Fair point. I'll still keep the verbose check in this file though because we need to see if the display name matches the name.

absternator commented on app/src/app/components/contents/packets/PacketHeader.tsx:
```diff
@@ -1,13 +1,27 @@
+        <p className="text-muted-foreground">{packetId}</p>
+      </div>
+    );
+  } else {
```
dont else :smile:

absternator commented on app/src/msw/handlers/packetGroupSummaryHandlers.ts:
```diff
@@ -1,11 +1,14 @@
-    return res(ctx.json(mockPacketGroupSummary));
+  rest.get(`${appConfig.apiUrl()}/packets/packetGroupSummaries`, (req, res, ctx) => {
+    const url = new URL(req.url)
+    if (url.searchParams.get("filterName")) {
```
nice!!!

absternator reviewed (commented):
Great work works like a charm!! just left some comments.


just some extras:
- display description should probably match font (text-muted)
- in general avoid multiple API calls in a component 
-

david-mears-2 commented on api/app/src/main/kotlin/packit/repository/PacketRepository.kt:
```diff
@@ -24,24 +24,26 @@ interface PacketRepository : JpaRepository<Packet, String>
         nativeQuery = true
     )
-    fun findPacketGroupSummaryByName(filterName: String): List<PacketGroupSummary>
+    fun getPacketGroupSummariesBySearchString(filterName: String): List<PacketGroupSummary>
```
The idea is that you can filter by name or display_name, so we shouldn't privelege 'name' by putting it in the function name. Do you prefer '...bySearchFilter'?

EmmaLRussell replied:
Or `getFilteredPackerGroupSummaries`..? `filterName` parameter could maybe just be `filter`?

EmmaLRussell commented on app/src/app/components/contents/PacketGroup/PacketGroup.tsx:
```diff
@@ -1,19 +1,48 @@
 export const PacketGroup = () => {
   const { packetName } = useParams();
+  const { packetGroupSummaries, error: packetGroupError } =
+    useGetPacketGroupSummaries(0, 1, packetName as string);
```
Is this just calling the search endpoint? It feels like it would be more efficient to have an endpoint which only gets a single packet group with the known name?

david-mears-2 replied:
There's no guarantee that there is a 'single' packet group with that name (also, the column isn't unique), so we have to do a search.

david-mears-2 replied:
Yes.
The new endpoint would need to also return the id of the latest packet, meaning its quite a lot of extra code just for a neater call here.

david-mears-2 replied:
Too late for the separate ticket I'm afraid. Added a new endpoint that gives back the single packet group's display name, along with the needed id of the latest packet.

EmmaLRussell reviewed (commented):
I like the styling on this! Seems pretty intuitive to me. 

> I have not made changes to the runner logs page, which is shown below, as it would be fairly in depth (requires either another request to get the packet metadata from outpack server, or a db query to get the packet display name) and I'm not sure it's needed. But I could do it in another PR if required.

It would definitely be one for another PR. Not convinced it's needed.

I feel like the text "Here's a list of all packets for the packet group 'Packet with description':" is not really needed, and is now just taking up space...

As discussed, maybe it would make sense to save the display name of the packet group on packet ingestion, from the latest packet per group. That might reduce the number of queries we need to do.

david-mears-2 commented:
> display description should probably match font (text-muted)

I will make resolve the inconsistency in the other direction:

![image](https://github.com/user-attachments/assets/9f9ef9a4-c903-4b41-8e62-8106f14909ec)

david-mears-2 commented on api/app/src/main/kotlin/packit/repository/PacketRepository.kt:
```diff
@@ -24,24 +24,26 @@ interface PacketRepository : JpaRepository<Packet, String>
                 ORDER BY start_time DESC
          """,
-        countQuery = "SELECT count(distinct name) from packet WHERE name ILIKE  %?1%",
+        countQuery = "SELECT count(distinct name) from packet WHERE name ILIKE %?1%",
```
Good point. I'm not sure what it's used for either - @absternator ?

david-mears-2 replied:
I see
> Defines a special count query that shall be used for pagination queries to lookup the total number of elements for a page. If none is configured we will derive the count query from the original query or countProjection() query if any.

So I'm inclined to just remove it if it falls back to the original query?

david-mears-2 commented on api/app/src/main/kotlin/packit/service/PacketService.kt:
```diff
@@ -44,14 +45,33 @@ class BasePacketService(
+        val orderlyDisplayName = description?.get("display") as? String
+        return if (!orderlyDisplayName.isNullOrBlank()) {
+            orderlyDisplayName
+        } else if (packet.custom?.get("display") is String) {
+            packet.custom["display"] as String
```
Ah yes - I forgot about the key that tells you which schema you're in. So I should in fact unit test this behaviour.

david-mears-2 commented on api/app/src/main/kotlin/packit/controllers/PacketGroupController.kt:
```diff
@@ -4,15 +4,17 @@ import org.springframework.data.domain.Page
 
 @Controller
-@RequestMapping("/packetGroup")
+@RequestMapping("/packetGroups")
```
by convention I think resources are plural in CRUD; in any case we already use the plural form for the packets controller.

david-mears-2 commented:
Main changes since last reviews:
* I added a 'latest_display_name' column to the packet **_group_** table, which is populated when packets are ingested.
* I added an endpoint (and various related gubbins like repository methods) that, when given the name of a packet group, provides its display name and the id of the latest packet with the same name. That was for use by PacketGroup.tsx and the benefit is in having SQL that can look up by name rather than rely on the LIKE function.

Neither of these allow us to have fewer network calls in the PacketGroup component.

EmmaLRussell commented on api/app/src/main/kotlin/packit/model/dto/PacketGroupDisplay.kt:
```diff
@@ -0,0 +1,11 @@
+package packit.model.dto
+
+interface PacketGroupDisplay {
+    val displayName: String
+    val packetDescription: String?
+}
+
+data class PacketGroupDisplayImpl(
+    override val displayName: String,
+    override val packetDescription: String?
+) : PacketGroupDisplay
```
We wouldn't usually bother with an interface for a data class .

EmmaLRussell commented on api/app/src/main/kotlin/packit/model/dto/PacketGroupDto.kt:
```diff
@@ -2,5 +2,6 @@ package packit.model.dto
 
 data class PacketGroupDto(
     val name: String,
+    val latestDisplayName: String,
```
I'm a little bit uneasy about the inconsistency between "displayName" for `PacketGroupDisplay` and "latestDisplayName" for `PacketGroupDto`. I guess there's potentially a subtle difference here which justifies the discrepancy in that `PacketGroupDisplay` is an ephemeral sort of thing so its display name is just the display name now, whereas `PacketGroup` persists and can have different "latestDisplayNames" over time..? But i think I'd still be inclined to just use the same prop for both to keep things simple.

david-mears-2 replied:
Renamed all to 'latest...'

EmmaLRussell commented on api/app/src/main/kotlin/packit/model/dto/PacketGroupDisplay.kt:
```diff
@@ -0,0 +1,11 @@
+
+interface PacketGroupDisplay {
+    val displayName: String
+    val packetDescription: String?
```
Is this the latest packet description, in the same way the display name is the latest packet display name?  I think just "description" would probably be a more consistent name. 

Also - should latest description be in the packet group table as well as latest display name?

david-mears-2 replied:
Renamed

EmmaLRussell commented on api/app/src/main/kotlin/packit/service/PacketGroupService.kt:
```diff
@@ -2,25 +2,42 @@ package packit.service
         return PagingHelper.convertListToPage(packetGroups, pageablePayload)
     }
+
+    override fun getPacketGroupDisplay(name: String): PacketGroupDisplay{
+        val latestPacketId = packetGroupRepository.findLatestPacketIdForGroup(name)?.id
+        val packetGroup = packetGroupRepository.findByName(name)
+        if (latestPacketId == null || packetGroup == null) {
+            throw PackitException("doesNotExist", HttpStatus.NOT_FOUND)
+        }
+        val packetOrderlyMetadata = packetService.getMetadataBy(latestPacketId).custom?.get("orderly") as? Map<*, *>
+        val packetLongDescription = (packetOrderlyMetadata?.get("description") as? Map<*, *>)?.get("long") as? String
+        return PacketGroupDisplayImpl(packetGroup.latestDisplayName, packetLongDescription)
+    }
```
I'm a bit confused by what's going on here- it's going to outpack server to fetch metadata at the time when the endpoint is called? I hadnt imagined that would happen, but rather that during `importPackets`. it would check for any new packet display names and update the packet group table accordingly - and then we'd return that display information when requested by front end. 

Oh... it looks like you are doing that. So this get from outpack server is just for descriptions? Why not just put those in the packet group table too?

david-mears-2 replied:
I guess I thought display names would be much more commonly referred to than descriptions, justifying the db column, but in practice that turned out not to be true.

EmmaLRussell commented on api/app/src/main/resources/db/migration/V20241206__add_latest_display_name_column_to_packet_group.sql:
```diff
@@ -0,0 +1,8 @@
+ALTER TABLE "packet_group"
+    ADD COLUMN "latest_display_name" text DEFAULT '';
+
+ALTER TABLE "packet_group"
+    ALTER COLUMN "latest_display_name" SET NOT NULL;
+
+ALTER TABLE "packet_group"
+    ALTER COLUMN "latest_display_name" DROP DEFAULT;
```
We always want the `latest_display_name`columns to be populated with non-empty values  don't we? So perhaps here we should also update any existing packet groups with display name. However if we were going to do that properly (not just copying over latest packet name) we'd need to get into the metadata, which would need to be done in code (getting metadata from outpack server). One for another ticket?

EmmaLRussell commented on app/src/app/components/contents/PacketGroup/hooks/useGetPacketPages.ts:
```diff
@@ -3,14 +3,16 @@ import appConfig from "../../../../../config/appConfig";
 import { PageablePackets } from "../../../../../types";
 
-export const useGetPacketGroups = (packetName: string | undefined, pageNumber: number, pageSize: number) => {
+export const useGetPacketPages = (packetName: string | undefined, pageNumber: number, pageSize: number) => {
```
..similarly here, would `useGetPacketPage` be slightly more accurate?

EmmaLRussell reviewed (commented):
Good stuff. 

Mostly nitpicks about naming. I think we might also need to think more about migration for existing packet groups, since there are some in the wild. And I'm wondering why we don't treat packet group description in the same way we do display name. 

It would be good to show the display name in the breadcrumb - in another ticket!
![image](https://github.com/user-attachments/assets/21321b40-e744-46b3-b554-0613c4885392)

I'd be in favour of doing any substantive changes in other tickets, as I feel like this branch has become quite a catch-all already, so it would be good to merge and then refine... though we should consider any deployment schedules if we do that (in context of migrations).

david-mears-2 commented:
> It would be good to show the display name in the breadcrumb - in another ticket!

Created: https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-6115/Use-the-packet-group-display-name-in-the-breadcrumbs

david-mears-2 commented:
I see two possible ways forward with regard to the two packet group (henceforth 'group') properties, display name and description, that are ultimately drawn from their latest packet.

1) Treat groups' descriptions like this branch currently treats display_names: add a db column for latest description, and add a script (in a future PR) for back-filling the new columns with reference to the metadata from outpack_server, run that script, and remove it.

2) Revert the db changes and have the /packetGroups/{name}/display endpoint query only the outpack_server for both display_name and description (as it currently does in this branch for description only); this prevents any complication / work about migrating existing groups using a script that queries outpack_server.

Here are all of the places when we'll want to refer in the front end to those properties of groups:
a) Breadcrumbs (display name)
b) Group page (display name and description)

(A bit counter-intuitively, the home page's table of groups makes no reference to packet_group db table, only packet db table.)

Other features we might bring in that relate to group display names:

i) Parameterised display names, e.g. (e.g., "diagnostic report for ${disease}"), as mentioned [here](https://resideglobal.slack.com/archives/C03FU2C9PMG/p1733485896428619?thread_ts=1733416423.989409&cid=C03FU2C9PMG). This would vary per packet in the group, and, in this situation, it would make less sense to privilege the latest packet over others; we might want to use the (constant) _name_ for the packet group instead of the display name. This reduces the need for a (latest_)display_name column; we'd probably want to get rid of that column in order to avoid confusing ourselves about the canonical display name for any group.

ii) On the packet group page, an info / tooltip warning that the packets' display names differ, as described in [the same thread](https://resideglobal.slack.com/archives/C03FU2C9PMG/p1733493113856579?thread_ts=1733416423.989409&cid=C03FU2C9PMG) and ticketised [here](https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-6117/Packet-group-page-warn-user-when-packets-display-names-differ). This implies that the packet group display page would need to check the display_name on each of its member packets anyway, obviating the savings achieved by putting a column on the group table.

There is a cost to using our proxy db less: more outward calls to outpack_server. Breadcrumbs would be rendered a little more slowly.

But overall I feel that the database columns don't gain us much, since the latest_display_name column would be used only in two components, and the description column only in one. It doesn't feel worth the added complexity of cacheing the info in the group table, potentially to unpick the code later when we try (i) or (ii).

Thoughts?

david-mears-2 commented:
Closing in favour of https://github.com/mrc-ide/packit/pull/150
